src/client/browserType.ts

```diff
@@ -126,48 +123,30 @@ export class BrowserType extends ChannelOwner<channels.BrowserTypeChannel, chann
     assert(optionsOrWsEndpoint.wsEndpoint, 'options.wsEndpoint is required');
     return this._connect(optionsOrWsEndpoint.wsEndpoint, optionsOrWsEndpoint);
   }
+
   async _connect(wsEndpoint: string, params: Partial<ConnectOptions> = {}): Promise<Browser> {
     const logger = params.logger;
-    const paramsHeaders = Object.assign({'User-Agent': getUserAgent()}, params.headers);
-    return this._wrapApiCall(async () => {
-      const ws = new WebSocket(wsEndpoint, [], {
-        perMessageDeflate: false,
-        maxPayload: 256 * 1024 * 1024, // 256Mb,
-        handshakeTimeout: this._timeoutSettings.timeout(params),
-        headers: paramsHeaders,
-      });
-      const connection = new Connection(() => ws.close());
-
-      // The 'ws' module in node sometimes sends us multiple messages in a single task.
-      const waitForNextTask = params.slowMo
-        ? (cb: () => any) => setTimeout(cb, params.slowMo)
-        : makeWaitForNextTask();
-      connection.onmessage = message => {
-        // Connection should handle all outgoing message in disconnected().
-        if (ws.readyState !== WebSocket.OPEN)
-          return;
-        ws.send(JSON.stringify(message));
-      };
-      ws.addEventListener('message', event => {
-        waitForNextTask(() => {
-          try {
-            // Since we may slow down the messages, but disconnect
-            // synchronously, we might come here with a message
-            // after disconnect.
-            if (!connection.isDisconnected())
-              connection.dispatch(JSON.parse(event.data));
-          } catch (e) {
-            console.error(`Playwright: Connection dispatch error`);
-            console.error(e);
-            ws.close();
-          }
-        });
+    return await this._wrapApiCall(async (channel: channels.BrowserTypeChannel) => {
+      const timeoutPromise = new ManualPromise<Browser>();
+      const timer = params.timeout ? setTimeout(() => timeoutPromise.reject(new Error(`Timeout ${params.timeout}ms exceeded.`)), params.timeout) : undefined;
+
+      const { pipe } = await channel.connect({ wsEndpoint, headers: params.headers, timeout: params.timeout });
+      const connection = new Connection(() => pipe.close().catch(() => {}));
+      connection.onmessage = message => pipe.send({ message }).catch(() => { });
+      pipe.on('message', ({ message }) => {
+        try {
+          // Since we may slow down the messages, but disconnect
```

Comment:
I think this is not relevant anymore.

Reply:
The comment isn't, but I think the check still is.